### PR TITLE
New: Site & User options API [ED-5573]

### PR DESCRIPTION
**Pull request by matipojo**

### CODE REVIEW

File: core/options/manager.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	public function get_name() {
+		return 'options';
+	}
+
+	/**
+	 * @param Option_Base|string $classname
+	 */
+	public function register( $classname ) {
```
Comment (StyleShit) [inline core/options/manager.php:28]: Add an `unregister()` method so it'll be consistent with other managers

File: core/options/manager.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+class Manager extends BaseModule {
+
+	private $registered = [];
+
+	public function __construct() {
+		parent::__construct();
+
+		do_action( 'elementor/options/register' );
```
Comment (StyleShit) [inline core/options/manager.php:18]:
```suggestion
		do_action( 'elementor/options/register', $this );
```

File: core/options/manager.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		return 'options';
+	}
+
+	/**
+	 * @param Option_Base|string $classname
+	 */
+	public function register( $classname ) {
+		$this->registered[ $classname::get_key() ] = $classname::get_key();
+
+		if ( method_exists( $classname, 'on_change' ) ) {
```
Comment (StyleShit) [inline core/options/manager.php:31]: Looks like `on_change` always exists, since it's defined in the base class

File: core/options/manager.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	/**
+	 * @param $key
+	 *
+	 * @return Option_Base
+	 */
+	public function get( $key ) {
+		if ( ! isset( $this->registered[ $key ] ) ) {
+			wp_die( esc_attr( $key ) . ' is not a valid option.' );
```
Comment (StyleShit) [inline core/options/manager.php:51]: Why not exception?

File: core/options/manager.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+		}
+
+		return $this->registered[ $key ];
```
Comment (StyleShit) [inline core/options/manager.php:54]: Maybe return all when a `$key` is not set?

File: core/options/user-option.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+	public function __construct( $user_id ) {
+		static::$user_id = $user_id;
+	}
+
+	public static function get() {
+		$value = get_user_option( static::get_full_key(), static::$user_id );
+		if ( false === $value ) {
+			$value = static::get_default();
+		}
+		return $value;
```
Comment (StyleShit) [inline core/options/user-option.php:21]:
```suggestion
		$value = get_user_option( static::get_full_key(), static::$user_id );
		
		return ( false === $value ) ? static::get_default() : $value;
```

Or at least add some space:
```suggestion
		$value = get_user_option( static::get_full_key(), static::$user_id );
		
		if ( false === $value ) {
			$value = static::get_default();
		}
		
		return $value;
```

File: includes/beta-testers.php [hunk 7]
```diff
@@ -19,6 +21,9 @@ class Beta_Testers {
 
 	const NEWSLETTER_PRIVACY_URL = 'https://go.elementor.com/beta-testers-newsletter-privacy';
 
+	/**
+	 * @deprcated Use User_Beta_Tester::SIGNUP
```
Comment (StyleShit) [inline includes/beta-testers.php:25]:
```suggestion
	 * @deprecated 3.5.0 Use `User_Beta_Tester::SIGNUP` instead.
```

Should also be added to the deprecations document.

File: includes/tracker.php [hunk 8]
```diff
@@ -60,7 +63,7 @@ public static function init() {
 	 * @return string Return `yes` if tracking allowed, `no` otherwise.
 	 */
 	public static function check_for_settings_optin( $new_value ) {
-		$old_value = get_option( 'elementor_allow_tracking', 'no' );
+		$old_value = Site_Usage_Opt_In::get();
 		if ( $old_value !== $new_value && 'yes' === $new_value ) {
```
Comment (StyleShit) [inline includes/tracker.php:67]:
```suggestion
		$old_value = Site_Usage_Opt_In::get();
		
		if ( $old_value !== $new_value && 'yes' === $new_value ) {
```

File: includes/tracker.php [hunk 9]
```diff
@@ -162,9 +165,11 @@ public static function send_tracking_data( $override = false ) {
 	 * @since 1.0.0
 	 * @access public
 	 * @static
+	 * @deprcated since 3.6.0
```
Comment (StyleShit) [inline includes/tracker.php:168]:
```suggestion
	 * @deprecated since 3.5.0
```

File: includes/tracker.php [hunk 10]
```diff
@@ -162,9 +165,11 @@ public static function send_tracking_data( $override = false ) {
...
 	 */
 	public static function is_allow_track() {
-		return 'yes' === get_option( 'elementor_allow_tracking', 'no' );
+		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( 'Tracker::is_allow_track', '3.6.0', 'Site_Usage_Opt_In::is_on()' );
+		return Site_Usage_Opt_In::is_on();
```
Comment (StyleShit) [inline includes/tracker.php:172]: Is it 3.5.0 or 3.6.0?

```suggestion
		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( 'Tracker::is_allow_track()', '3.6.0', 'Site_Usage_Opt_In::is_on()' );
		return Site_Usage_Opt_In::is_on();
```

File: includes/user.php [hunk 11]
```diff
@@ -22,8 +24,14 @@ class User {
 	 */
 	const ADMIN_NOTICES_KEY = 'elementor_admin_notices';
 
+	/**
+	 * @deprcated use User_Introduction
```
Comment (StyleShit) [inline includes/user.php:28]: Also in other places.

```suggestion
	 * @deprecated 3.5.0 use `User_Introduction::get_key()` instead.
```

File: tests/phpunit/elementor/includes/settings/test-settings.php [hunk 12]
```diff
@@ -1,12 +1,17 @@
...
+use Elementor\Core\Admin\Options\Site_CSS_Print_Method;
+use Elementor\Plugin;
 use ElementorEditorTesting\Elementor_Test_Base;
 use Elementor\Core\Files\CSS\Global_CSS;
 
 class Test_Settings extends Elementor_Test_Base {
 
 	public function test_clear_css_cache_on_update_css_settings() {
+		// Register the option in order to listen to the changes.
+		Plugin::$instance->options->register( Site_CSS_Print_Method::class );
```
Comment (StyleShit) [inline tests/phpunit/elementor/includes/settings/test-settings.php:13]: Isn't it registered by default?

https://github.com/elementor/elementor/blob/ae434047f20f115377d41b957e7c73040bf3962c/core/admin/admin.php#L687

File: tests/phpunit/elementor/includes/test-tracker.php [hunk 13]
```diff
@@ -40,7 +42,7 @@ public function test_get_settings_advanced_usage() {
 		// Load font_awesome_support settings.
 		Plugin::$instance->icons_manager->register_admin_settings( Plugin::$instance->settings );
 
-		update_option( 'elementor_css_print_method', 'internal' );
+		Site_CSS_Print_Method::set( 'internal' );
```
Comment (StyleShit) [inline tests/phpunit/elementor/includes/test-tracker.php:45]: Maybe that could be useful? Like `set_on()` / `set_off()`

```suggestion
		Site_CSS_Print_Method::set_internal();
```

File: core/options/site-option.php [hunk 14]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Elementor\Core\Options;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+abstract class Site_Option extends Option_Base {
```
Comment (StyleShit) [inline core/options/site-option.php:9]: Tests?

File: core/options/option-base.php [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+namespace Elementor\Core\Options;
+
+use Elementor\Core\Admin\Options\User_Introduction;
+use Elementor\Plugin;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+abstract class Option_Base {
```
Comment (StyleShit) [inline core/options/option-base.php:12]: Tests?

**Review by StyleShit — CHANGES_REQUESTED** [review]
Great work! 😎

File: core/admin/options/site-css-print-method.php [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+	exit; // Exit if accessed directly
+}
+
+class Site_CSS_Print_Method extends Site_Option {
+	public static function get_key() {
+		return 'css_print_method';
+	}
+
+	public static function get_default() {
+		return 'external';
```
[the remark this answers is not in the document]
Comment (matipojo) [reply core/admin/options/site-css-print-method.php]: Done

File: core/admin/options/site-usage-notice.php [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+	public static function is_off() {
+		return '1' === static::get();
+	}
+
+	public static function set_on() {
+		return static::delete();
+	}
+
+	public static function set_off() {
+		return static::set( 1 );
```
[the remark this answers is not in the document]
Comment (matipojo) [reply core/admin/options/site-usage-notice.php:36]: It's saved as a string.

File: core/options/manager.php (hunk 1 above)
Comment (matipojo) [reply core/options/manager.php:28]: In this case we won't the ability to unregister

File: core/options/manager.php (hunk 5 above)
Comment (matipojo) [reply core/options/manager.php:55]: I added a `get_all` method

File: tests/phpunit/elementor/includes/settings/test-settings.php (hunk 12 above)
Comment (matipojo) [reply tests/phpunit/elementor/includes/settings/test-settings.php:13]: It happens on the `admin.php`, but it seems that the test is running in another context

File: includes/tracker.php [hunk 18]
```diff
@@ -170,9 +174,11 @@ public static function send_tracking_data( $override = false ) {
 	 * @since 1.0.0
 	 * @access public
 	 * @static
+	 * @deprcated since 3.6.0
 	 */
 	public static function is_allow_track() {
-		return 'yes' === get_option( 'elementor_allow_tracking', 'no' );
+		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( 'Tracker::is_allow_track', '3.6.0', 'Site_Usage_Opt_In::is_on()' );
```
Comment (matipojo) [inline includes/tracker.php:180]:
```suggestion
		Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function( 'Tracker::is_allow_track', '3.14.0', 'Site_Usage_Opt_In::is_on()' );
```

File: includes/user.php [hunk 19]
```diff
@@ -22,8 +24,14 @@ class User {
 	 */
 	const ADMIN_NOTICES_KEY = 'elementor_admin_notices';
 
+	/**
+	 * @deprecated 3.6.0 use `User_Introduction::get_full_key()` instead.
```
Comment (matipojo) [inline includes/user.php:28]:
```suggestion
	 * @deprecated 3.14.0 use `User_Introduction::get_full_key()` instead.
```

File: includes/user.php [hunk 20]
```diff
@@ -22,8 +24,14 @@ class User {
...
+	 */
 	const INTRODUCTION_KEY = 'elementor_introduction';
 
+	/**
+	 * @deprecated 3.6.0 use `Site_Beta::get_full_key()` instead.
```
Comment (matipojo) [inline includes/user.php:33]:
```suggestion
	 * @deprecated 3.14.0 use `Site_Beta::get_full_key()` instead.
```

**Comment by arielk** [comment]
Is this PR still relevant?

File: core/admin/options/site-css-print-method.php [hunk 21]
```diff
@@ -16,6 +16,13 @@ public static function get_key() {
 		return 'css_print_method';
 	}
 
+	public static function get_options(){
+		return [
+			static::OPTION_EXTERNAL => __( 'External File', 'elementor' ),
+			static::OPTION_INTERNAL => __( 'Internal Embedding', 'elementor' ),
```
Comment (rodolphebertozzo) [inline core/admin/options/site-css-print-method.php:22]: esc

File: core/options/boolean-option-base.php [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+abstract class Boolean_Option_Base extends Option_Base {
+	const OPTION_TRUE = true;
+	const OPTION_FALSE = false;
+
+	protected static function get_options() {
+		return [
+			static::OPTION_TRUE => __( 'True', 'elementor' ),
+			static::OPTION_FALSE => __( 'False', 'elementor' ),
```
Comment (rodolphebertozzo) [inline core/options/boolean-option-base.php:16]: esc

File: core/options/yes-no-option-base.php [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+abstract class Yes_No_Option_Base extends Boolean_Option_Base {
+
+	protected static function get_options() {
+		return [
+			static::OPTION_TRUE => __( 'Yes', 'elementor' ),
+			static::OPTION_FALSE => __( 'No', 'elementor' ),
```
Comment (rodolphebertozzo) [inline core/options/yes-no-option-base.php:14]: esc

**Review by rodolphebertozzo — CHANGES_REQUESTED** [review]

File: core/admin/config/site-usage-last-send.php [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+	public static function get_default() {
+		return false;
+	}
+
+	public static function should_autoload() {
+		return false;
+	}
+
+	protected static function validate($value) {
+		return is_int($value);
```
Comment (matipojo) [inline core/admin/config/site-usage-last-send.php:24]:
```suggestion
		return is_int( $value );
```

File: core/admin/config/wp-blog-description.php [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+	public static function get_key() {
+		return 'blogdescription';
+	}
+
+	public static function get_default() {
+		return '';
+	}
+
+	protected static function validate($value) {
+		return is_scalar($value);
```
Comment (matipojo) [inline core/admin/config/wp-blog-description.php:20]:
```suggestion
		return is_scalar( $value );
```

File: core/admin/config/wp-blog-name.php [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+	public static function get_key() {
+		return 'blogname';
+	}
+
+	public static function get_default() {
+		return '';
+	}
+
+	protected static function validate($value) {
+		return !empty($value) && is_string($value);
```
Comment (matipojo) [inline core/admin/config/wp-blog-name.php:20]:
```suggestion
		return !empty( $value ) && is_string( $value );
```

File: core/admin/config/wp-home.php [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+	public static function get_key() {
+		return 'home';
+	}
+
+	public static function get_default() {
+		return '';
+	}
+
+	protected static function validate($value) {
+		return !empty($value) && is_string($value);
```
Comment (matipojo) [inline core/admin/config/wp-home.php:20]:
```suggestion
		return !empty( $value ) && is_string( $value );
```

File: core/admin/config/wp-site-icon.php [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+	public static function get_key() {
+		return 'site_icon';
+	}
+
+	public static function get_default() {
+		return '';
+	}
+
+	protected static function validate($value) {
+		return is_scalar($value);
```
Comment (matipojo) [inline core/admin/config/wp-site-icon.php:20]:
```suggestion
		return is_scalar( $value );
```

File: core/api/config/site-remote-feed.php [hunk 29]
```diff
@@ -0,0 +1,27 @@
...
+	public static function get_default() {
+		return [];
+	}
+
+	public static function should_autoload() {
+		return false;
+	}
+
+	protected static function validate($value) {
+		return is_array($value);
```
Comment (matipojo) [inline core/api/config/site-remote-feed.php:25]:
```suggestion
		return is_array( $value );
```

File: core/api/config/site-remote-library.php [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+	public static function get_default() {
+		return [];
+	}
+
+	public static function should_autoload() {
+		return false;
+	}
+
+	protected static function validate($value) {
+		return is_array($value);
```
Comment (matipojo) [inline core/api/config/site-remote-library.php:24]:
```suggestion
		return is_array( $value );
```

File: core/admin/admin.php [hunk 31]
```diff
@@ -839,6 +846,17 @@ public function init_beta_tester( $current_screen ) {
...
+		Plugin::$instance->config->register( Site_Beta::class );
+		Plugin::$instance->config->register( Site_CSS_Print_Method::class );
+		Plugin::$instance->config->register( Site_Font_Display::class );
+		Plugin::$instance->config->register( Site_Google_Fonts::class );
+		Plugin::$instance->config->register( Site_Usage_Last_Send::class );
+		Plugin::$instance->config->register( Site_Usage_Notice::class );
+		Plugin::$instance->config->register( Site_Usage_Opt_In::class );
+		Plugin::$instance->config->register( User_Beta_Tester::class );
+		Plugin::$instance->config->register( User_Introduction::class );
+		Plugin::$instance->config->register( WP_Blog_Name::class );
```
Comment (matipojo) [inline core/admin/admin.php:858]: Remove

File: core/config/config-base.php [hunk 32]
```diff
@@ -0,0 +1,143 @@
...
+	exit; // Exit if accessed directly
+}
+
+abstract class Config_Base {
+	const PREFIX = '';
+
+	const VALUE_TRUE = true;
+	const VALUE_FALSE = false;
+
+	public static function get_options() {
```
Comment (matipojo) [inline core/config/config-base.php:17]: Abstract, Move to a trait

File: core/config/config-base.php [hunk 33]
```diff
@@ -0,0 +1,143 @@
...
+
+	/**
+	 * @param mixed $value
+	 *
+	 * @return bool
+	 */
+	final public static function set( $value ) {
+		$old_value = static::get();
+
+		if ( ! Manager::is_admin() ) {
```
Comment (matipojo) [inline core/config/config-base.php:50]:
```suggestion
		if ( ! Manager::is_backend() ) {
```

File: core/config/config-base.php [hunk 34]
```diff
@@ -0,0 +1,143 @@
...
+			throw new \Error( 'Config can only be changed in the admin' );
+		}
+
+		// Avoid changing to a value that is not in the options.
+		if ( ! static::validate( $value ) ) {
+			throw new \Error( static::class . ': Invalid value: ' . var_export( $value, true ) );
+		}
```
Comment (matipojo) [inline core/config/config-base.php:57]: Move to trait

File: core/config/site-config-base.php [hunk 35]
```diff
@@ -0,0 +1,22 @@
...
+
+abstract class Site_Config_Base extends WP_Option_Base {
+	const PREFIX = 'elementor_';
+
+	/**
+	 * @return bool
+	 */
+	public static function should_autoload() {
+		throw new \Error( __METHOD__ . ' must be implemented' );
+	}
```
Comment (matipojo) [inline core/config/site-config-base.php:17]: Remove

File: core/config/config-base.php [hunk 36]
```diff
@@ -0,0 +1,143 @@
...
+		return static::deleter();
+	}
+
+	public static function get_prefix() {
+		return static::PREFIX;
+	}
+
+	public static function get_full_key() {
+		return static::get_prefix() . static::get_key();
+	}
```
[the remark this answers is not in the document]
Comment (matipojo) [reply core/config/config-base.php:83]: Changed to an abstract method: `get_db_key`

**Pull request closed without merging**